File: frontend/solutions/p7/aiSolution.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>();
        for (int a : asteroids) {
            boolean alive = true;
            while (alive && a < 0 && !stack.isEmpty() && stack.peekLast() > 0) {
                int top = stack.peekLast();
                if (top < -a) {
                    stack.pollLast();
                    continue;
                } else if (top == -a) {
                    stack.pollLast();
                }
                alive = false;
            }
            if (alive) stack.addLast(a);
        }

        int[] result = new int[stack.size()];
        int i = 0;
        for (int val : stack) result[i++] = val;
        return result;
    }
}

File: frontend/solutions/p7/aiSolution.js
// correct implementation stack data structure
// /**
//  * @param {number[]} asteroids
//  * @return {number[]}
//  */
// var asteroidCollision = function(asteroids) {
//     const stack = [];
//     for (const a of asteroids) {
//         let alive = true;
//         while (alive && a < 0 && stack.length && stack[stack.length - 1] > 0) {
//             const top = stack[stack.length - 1];
//             if (top < -a) {
//                 stack.pop();
//                 continue;
//             } else if (top === -a) {
//                 stack.pop();
//             }
//             alive = false;
//         }
//         if (alive) stack.push(a);
//     }
//     return stack;
// };

// contains synthetic error - uses queue instead of stack
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
    const queue = [];
    for (const a of asteroids) {
        let alive = true;
        while (alive && a < 0 && queue.length && queue[0] > 0) {
            const front = queue[0];
            if (front < -a) {
                queue.shift();
                continue;
            } else if (front === -a) {
                queue.shift();
            }
            alive = false;
        }
        if (alive) queue.push(a);
    }
    return queue;
};